[cli][smoke-test] Make genesis and stake configurable in swarm and test e2e validator runbook

ikabiljo opened the pull request:
### Description

- Made genesis and stake configurable from local swarm tests
- In order to do that nicely, without every test needing to set everything, add SwarmBuilder. 
- removed LocalSwarmBuilder that was used always with same configuration

### Test Plan

Unit test added follows the full validator e2e runbook  from AIT2 :
https://aptos.dev/nodes/ait/connect-to-testnet#joining-validator-set
https://aptos.dev/nodes/ait/connect-to-testnet/#leaving-validator-set

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/2273)
<!-- Reviewable:end -->

zekun000 commented on aptos-node/src/lib.rs:
```diff
@@ -251,15 +251,18 @@ where
+            .with_init_genesis_config(Some(Arc::new(|genesis_config| {
+                genesis_config.allow_new_validators = true;
+                genesis_config.epoch_duration_secs = EPOCH_LENGTH_SECS;
+                genesis_config.initial_lockup_duration_secs = EPOCH_LENGTH_SECS;
```
is it intentional to remove the now_secs?

ikabiljo replied:
for easy configuration I changed config to have initial_lockup_duration_secs, and add now_secs when creating initial_lockup_timestamp for the actual genesis.

zekun000 commented on api/src/tests/test_context.rs:
```diff
@@ -57,10 +57,13 @@ pub fn new_test_context(test_name: &'static str) -> TestContext {
+        genesis_config.min_price_per_gas_unit = 0;
+        genesis_config.min_lockup_duration_secs = 0;
+        genesis_config.max_lockup_duration_secs = 86400;
+        // genesis_config.initial_lockup_duration_secs;
```
what's this comment for?

ikabiljo replied:
this is the only place that config logically changed. previously it used initial_lockup_timestamp = 0, which cannot be specified any more (as we use duration after current timestamp - so initial_lockup_timestamp cannot be before NOW).
If all tests pass, I'll make a comment that that was changed, otherwise I'll need to see to work around it.

ikabiljo commented:
This test just caught the breakage in widthdraw stake CLI, after "amount" was added to the move code :)

updated PR with the fix

movekevin commented on testsuite/smoke-test/src/aptos_cli.rs:
```diff
@@ -211,6 +169,185 @@ async fn test_register_and_update_validator() {
+
+    assert_validator_set_sizes(&cli, validator_cli_index, 1, 0, 0).await;
+
+    gas_used += get_gas(cli.join_validator_set(validator_cli_index).await.unwrap());
```
Do we want to test that the validator can now vote after joining the validator set?

ikabiljo replied:
I am not actually creating a validator node. I am having some issues with private keys matching. 

I think this test covers enough as it is, and we can have a separate test for that.

movekevin commented on testsuite/smoke-test/src/aptos_cli.rs:
```diff
@@ -211,6 +169,185 @@ async fn test_register_and_update_validator() {
+    let lockup_left = lockup_duration
+        .as_secs()
+        .saturating_sub(lockup_start.elapsed().unwrap().as_secs());
+    println!("Lockup left: {}", lockup_left);
```
Remove println!

ikabiljo commented on api/src/tests/test_context.rs:
```diff
@@ -57,10 +57,13 @@ pub fn new_test_context(test_name: &'static str) -> TestContext {
-    .with_min_lockup_duration_secs(0)
-    .with_max_lockup_duration_secs(86400)
-    .with_initial_lockup_timestamp(0)
+    // not sure if this is intentional or needed?
```
I am not sure if these overrides are important for the test, or if we just put them to initialize the builder?

i.e. do we want to have values here different from the default?

but yeah, this comment was for the review, I'll remove the comment.

gregnazario commented on crates/aptos-genesis/src/builder.rs:
```diff
@@ -489,9 +451,10 @@ impl Builder {
         // Generate root key
         let root_key = keygen.generate_ed25519_private_key();
 
+        let template = NodeConfig::default_for_validator();
```
Why this change, now you can't override the default configuration?

gregnazario commented on crates/aptos-genesis/src/builder.rs:
```diff
@@ -421,53 +420,16 @@ impl Builder {
         self
     }
 
-    pub fn with_template(mut self, template: NodeConfig) -> Self {
-        self.template = template;
-        self
-    }
-
```
This is used in test mode for a local testnet (or at least it's supposed to be)

gregnazario reviewed (changes requested):
Genesis builder is used for local test mode, so it should take in a template that isn't necessarily the default.  Unless I'm missing something?

ikabiljo commented on aptos-node/src/lib.rs:
```diff
@@ -251,15 +251,18 @@ where
-            .with_min_lockup_duration_secs(0)
-            .with_max_lockup_duration_secs(86400 * 14);
+            .with_init_config(Some(Arc::new(move |_, config, _| {
+                *config = template.clone();
```
Here we copy the template into config, as before

ikabiljo reviewed (commented):
@gregnazario - I just consolidated two ways for setting configuration - template and init_config, by removing one.

The template is still used and set correctly for your usecase, it is just set through with_init_config (which is more flexible)
see the inline comment:

ikabiljo commented:
Rebased on top of -  https://github.com/aptos-labs/aptos-core/pull/2336.

Added some more changes to that - passes operator args where needed (otherwise commands fatal), creating add_cli_account utility to wrap adding cli accounts.

gregnazario commented on crates/aptos/src/test/mod.rs:
```diff
@@ -274,13 +291,17 @@ impl CliTestFramework {
         .await
     }
 
-    pub async fn init(&self, private_key: &Ed25519PrivateKey) -> CliTypedResult<()> {
+    pub async fn init(
+        &self,
+        profile: String,
+        private_key: &Ed25519PrivateKey,
+    ) -> CliTypedResult<()> {
         InitTool {
             rest_url: Some(self.endpoint.clone()),
             faucet_url: Some(self.faucet_endpoint.clone()),
             rng_args: RngArgs::from_seed([0; 32]),
             private_key_options: PrivateKeyInputOptions::from_private_key(private_key)?,
-            profile_options: Default::default(),
+            profile_options: ProfileOptions { profile },
```
Remove the profiles, they were a cause of flakiness in the tests.  Should be back to default.

ikabiljo replied:
this was actually intentional, but let me leave that for a separate diff.

ikabiljo commented:
for why configuring with template is so complicated, we want to:

  -  have defaults
  -  have people override defaults they want
  - have framework adjust configs (especially for NodeConfig appropriate for the environment)

If we only allow configuring based on template, there is no good way for smoke tests to set defaults. So we need to modify the config after default config is created - so I use lambda for that.
So if someone wants to override full config, they can copy it over - it just needs to happen in a lambda. your usecase is the only one that does that - all others modify just a few configs - for which new way makes more sense.




So I went with user passing in the init_config function to modify the configs, so framework does:
```
let config = get_default()
init_config(config)
configure framework related thingsand that gives us a full config we can store and use for nodes.
```
